### PR TITLE
bugfix internal/file and add CreateTemp function and resolve go module failure

### PR DESCRIPTION
**Pull request by kpango**
Signed-off-by: kpango <kpango@vdaas.org>

### Description:
This PR inclued 3 bugfix and 2 feature and deps update

Bugfix
1. internal/file.writeFile validates file existing in Sidecar case, it should be ignored when file error is not AlreadyExists
2. pkg/agent/ngt/service on CoW mode never checks non-CoW data, which is important for backwards compatibility
3. Go module failure detected, I resolved jwt-go problem by updating cobra and viper

Features
- Add CreateTemp function using MkTempDir for S3 future usage. 
- Remove ioutil package and move them to io or file package

Deps
- Update libs

### Environment:

- Go Version: 1.18
- Docker Version: 20.10.8
- Kubernetes Version: 1.22.0
- NGT Version: 1.14.3

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pkg/agent/core/ngt/service/ngt.go [hunk 1]
```diff
@@ -456,6 +459,12 @@ func (n *ngt) loadKVS(path string) (err error) {
 		return err
 	}
 
+	if n.kvs == nil {
+		n.kvs = kvs.New(kvs.WithConcurrency(n.kvsdbConcurrency))
+	} else if n.kvs.Len() > 0 {
+		n.kvs.Close()
+		n.kvs = kvs.New(kvs.WithConcurrency(n.kvsdbConcurrency))
```
Comment (vankichi) [inline pkg/agent/core/ngt/service/ngt.go:466]: Why does existing KVS trash in this moment?

File: pkg/agent/core/ngt/service/ngt.go [hunk 2]
```diff
@@ -427,13 +351,92 @@ func (n *ngt) initNGT(opts ...core.Option) (err error) {
...
+				"load vald secondary index success from %s\t error: %v\tbut index data are too few %d compared to primary data now trying to load from primary index data again from %s and start up with them",
+				n.oldPath,
+				err,
+				n.kvs.Len(),
+				n.oldPath,
+			)
+			err = n.load(ctx, n.path, opts...)
+			if err == nil {
+				return nil
+			}
```
Comment (vankichi) [inline pkg/agent/core/ngt/service/ngt.go:402]: I suppose, if err is not nil, should we output another log in the below line?
There is no chance to output this err in current implementation.

File: pkg/agent/core/ngt/service/ngt.go [hunk 3]
```diff
@@ -128,6 +130,8 @@ type ngt struct {
 
 	idelay time.Duration // initial delay duration
 	dcd    bool          // disable commit daemon
+
+	kvsdbConcurrency int // kvsdb concurrency
```
Comment (vankichi) [inline pkg/agent/core/ngt/service/ngt.go:134]: I could not understand what is the `kvsdbConcurrency` and what does it work for.
could you please ELI5?

File: pkg/agent/core/ngt/service/ngt.go (hunk 1 above)
Comment (kpango) [reply pkg/agent/core/ngt/service/ngt.go:466]: The loadKVS function is called at startup. In CoW, loadKVS is called in the order of origin, old, base, and tmp, but for each load, the previous load result may remain in kvs, so it is flushed.

File: pkg/agent/core/ngt/service/ngt.go (hunk 2 above)
Comment (kpango) [reply pkg/agent/core/ngt/service/ngt.go:402]: it would be shown at L407

File: pkg/agent/core/ngt/service/ngt.go (hunk 3 above)
Comment (kpango) [reply pkg/agent/core/ngt/service/ngt.go:134]: kvsdb uses a HashTable in Vald when processing all cases, etc., so it is possible to search in parallel in each Hash space (512 bits).
It is possible to search up to 512 threads, but generating many goroutines is costly, so it is made so that the degree of concurrency can be limited from the configuration.

File: pkg/agent/core/ngt/service/ngt.go (hunk 2 above)
Comment (vankichi) [reply pkg/agent/core/ngt/service/ngt.go:414]: my eyes were 🐸

**Review by vankichi — APPROVED** [review]
LTGM:smile:

**Review by hlts2 — DISMISSED** [review]

File: internal/file/file.go [hunk 4]
```diff
@@ -214,11 +223,14 @@ func CopyFileWithPerm(ctx context.Context, src, dst string, perm fs.FileMode) (n
...
 	switch {
-	case !exist || fi == nil || fi.Size() == 0:
+	case !exist, fi == nil, fi.Size() == 0, fi.IsDir():
 		return 0, errors.Wrap(err, errors.ErrFileNotFound(src).Error())
-	case err != nil && !os.IsExist(err):
+	case err != nil && (!errors.Is(err, fs.ErrExist) || errors.Is(err, fs.ErrNotExist)):
 		return 0, errors.Wrap(errors.ErrFileNotFound(src), err.Error())
+	case err != nil:
+		log.Warn(err)
 	}
```
Comment (kevindiu) [inline internal/file/file.go:232]: Is there any reason we use `switch` here?
I think we should check the `err` before checking `exist`?

File: internal/file/file.go [hunk 5]
```diff
@@ -264,14 +276,17 @@ func writeFile(ctx context.Context, target string, r io.Reader, flg int, perm fs
...
 	switch {
-	case err == nil, exist, fi != nil && fi.Size() != 0:
+	case err == nil, exist, fi != nil && fi.Size() != 0, fi != nil && fi.IsDir():
 		err = errors.ErrFileAlreadyExists(target)
-	case err != nil && !os.IsNotExist(err):
+	case err != nil && !errors.Is(err, fs.ErrNotExist), err != nil && errors.Is(err, fs.ErrExist):
 		err = errors.Wrap(errors.ErrFileAlreadyExists(target), err.Error())
+	case err != nil && !errors.Is(err, fs.ErrNotExist):
+		log.Warn(err)
 	}
```
Comment (kevindiu) [inline internal/file/file.go:285]: same as 
https://github.com/vdaas/vald/pull/1608/files#r845728577

File: internal/file/file.go (hunk 4 above)
Comment (kpango) [reply internal/file/file.go:232]: you prefer if branch?

Comment (kpango) [reply internal/file/file.go:232]: I think check order is not important in this case.

File: pkg/agent/core/ngt/service/ngt.go [hunk 6]
```diff
@@ -206,98 +217,64 @@ func New(cfg *config.NGT, opts ...Option) (nn NGT, err error) {
...
-				tpath := n.tmpPath.Load().(string)
-				log.Debugf("failed to find any index from %s and %s and %s now start with temporary directory as a new index path: %s", n.basePath, n.path, n.oldPath, tpath)
-				n.core, err = core.New(append(opts, core.WithIndexPath(tpath))...)
-				return err
-			}
-			path = n.basePath
-		} else {
-			path = n.oldPath
+		return err
+	case exist && fi != nil && fi.IsDir():
```
Comment (kevindiu) [inline pkg/agent/core/ngt/service/ngt.go:231]: same as https://github.com/vdaas/vald/pull/1608/files#r845728577

File: pkg/agent/core/ngt/service/ngt.go [hunk 7]
```diff
@@ -206,98 +217,64 @@ func New(cfg *config.NGT, opts ...Option) (nn NGT, err error) {
...
+	metadataPath := file.Join(path, metadata.AgentMetadataFileName)
+	log.Debugf("index path: %s exists, now starting to check metadata from %s", path, metadataPath)
+	exist, fi, err = file.ExistsWithDetail(metadataPath)
+	switch {
+	case !exist, fi == nil, fi != nil && fi.Size() == 0, err != nil && errors.Is(err, fs.ErrNotExist):
+		err = errors.Wrapf(errors.ErrIndexFileNotFound, "metadata file does not exists,\tpath: %s,\terr: %v", metadataPath, err)
+		return err
+	case err != nil && errors.Is(err, fs.ErrPermission):
+		if fi != nil {
+			err = errors.Wrap(errors.ErrFailedToOpenFile(err, metadataPath, 0, fi.Mode()), "invalid permission for loading metadata")
```
Comment (kevindiu) [inline pkg/agent/core/ngt/service/ngt.go:253]: same https://github.com/vdaas/vald/pull/1608/files#r845728577

File: internal/file/file.go (hunk 4 above)
Comment (kevindiu) [reply internal/file/file.go:232]: I think both `switch` or `if` branch is fine, but seems using `if` here will reduce the complexity of the checking.

```go
if err != nil {
   log.Warn(err) // should we put before checking the error?
  if (!errors.Is(err, fs.ErrExist) || errors.Is(err, fs.ErrNotExist)) {
      return 0, errors.Wrap(errors.ErrFileNotFound(src), err.Error())
   }
}

if !exist, fi == nil || fi.Size() == 0 || fi.IsDir() {
   return 0, errors.Wrap(err, errors.ErrFileNotFound(src).Error())
}

File: internal/compress/gob.go [hunk 8]
```diff
@@ -19,9 +19,10 @@ package compress
 
 import (
 	"bytes"
-	"io"
 	"reflect"
 
+	"github.com/vdaas/vald/internal/io"
+
```
Comment (kmrmt) [inline internal/compress/gob.go:25]:
```suggestion
```

File: internal/compress/gzip/gzip.go [hunk 9]
```diff
@@ -16,7 +16,7 @@
 package gzip
 
 import (
-	"io"
+	"github.com/vdaas/vald/internal/io"
```
Comment (kmrmt) [inline internal/compress/gzip/gzip.go:20]:
```suggestion
```

**Review by kmrmt — CHANGES_REQUESTED** [review]
empty line

**Review by kmrmt — DISMISSED** [review]

**Review by vankichi — APPROVED** [review]

**Review by hlts2 — APPROVED** [review]

**Review by kevindiu — DISMISSED** [review]

**Review by vankichi — APPROVED** [review]
